# Исследование надежности заемщиков


# Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов. 

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<b>Описание данных: \
children — количество детей в семье \
days_employed — общий трудовой стаж в днях \
dob_years — возраст клиента в годах \
education — уровень образования клиента \
education_id — идентификатор уровня образования \
family_status — семейное положение \
family_status_id — идентификатор семейного положения \
gender — пол клиента \
income_type — тип занятости \
debt — имел ли задолженность по возврату кредитов \
total_income — ежемесячный доход \
purpose — цель получения кредита
    
Существуют пропуски в стобцах 'days_employed' и 'total_income' нужно определить какого рода эти пропуски и заполнить их.\
В столбце 'days_employed' встречаются положительные и отрицательные значения. Возможно это связано с ошибкой при обработки данных из раных источников. Нужно привести все значения в единый формат для последующего анализа.


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = data.loc[(data['income_type'] == t), 'total_income'].median()


### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & 
             (data['days_employed'].isna()),'days_employed'] = data.loc[(data['income_type'] == t), 'days_employed'].median()


**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype('int32')

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

54

In [16]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [17]:
data['education'] = data['education'].str.lower()

In [18]:
data.duplicated().sum()

17

In [19]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  
0            253875                 покупка жилья                     B  
1            112080       приобретение автомобиля                     C  
2            145885                 покупка жилья                     C  
3            267628    дополнительное образование                     B  
4            158616               сыграть свадьбу                     C  
...             ...                           ...                   ...  
21520        224791             операции с жильем                     B  
21521        155999          сделка с автомобилем                     C  
21522         89672                  недвижимость                     C  
21523        244093  на покупку своего автомобиля                     B  
21524         82047         на покупку автомобиля                     C  

[21331 rows x 13 columns]

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [22]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Проверка какие бывают значения в столбце `'debt'`

In [25]:
data['debt'].unique() 

array([0, 1], dtype=int64)

Количество просрочек по отношению к детям

In [26]:
data.groupby('children')['debt'].sum() 

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

Общее количество займов по отношению к детям

In [27]:
data.groupby('children')['debt'].count()

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

In [28]:
(data.groupby('children')['debt'].sum() / data.groupby('children')['debt'].count()).sort_values(ascending=False)

children
4    0.097561
2    0.094542
1    0.092346
3    0.081818
0    0.075438
5    0.000000
Name: debt, dtype: float64

Сотношение просрочек к общему количеству займов
В выборке количество многодетных заемщиков с разным количеством детей достаточно мало и нет возможности сравнить количество просрочек по отношению количесва детей. Обьеденим многодетных заемщиков в одну группу и сравним их кредитоспособность с заемщиками без детей и заемщиками с одним ребенком.

In [29]:
def children_group(nomber_of_childrens):
    try:
        if nomber_of_childrens == 0:
            return 'нет детей'
        elif nomber_of_childrens == 1:
            return 'один ребенок'
        elif nomber_of_childrens > 1:
            return 'многодетная семья'
    except:
        return 'неверное значение'

In [30]:

data['presence_of_children'] = data['children'].apply(children_group)

(data.groupby('presence_of_children')['debt'].sum() / data.groupby('presence_of_children')['debt'].count()).sort_values(ascending=False)

presence_of_children
многодетная семья    0.092516
один ребенок         0.092346
нет детей            0.075438
Name: debt, dtype: float64

**Вывод: Заемщики не имеющие детей допускают меньше просрочек. Люди с одиним или более детей допускают одинаковое количество просрочек** 

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Проверка какие бывают значения в столбце `'family_status'`

In [31]:
data['family_status'].unique() 

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Количество просрочек по отношению к семейному статусу

In [32]:
data.groupby('family_status')['debt'].sum().sort_values(ascending=False)

family_status
женат / замужем          927
гражданский брак         385
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
Name: debt, dtype: int64

Общее количество займов по отношению к семейному статусу

In [33]:
data.groupby('family_status')['debt'].count().sort_values(ascending=False)

family_status
женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: debt, dtype: int64

Сотношение просрочек к общему количеству займов

In [34]:
(data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count()).sort_values(ascending=False)

family_status
Не женат / не замужем    0.097639
гражданский брак         0.093130
женат / замужем          0.075606
в разводе                0.070648
вдовец / вдова           0.066246
Name: debt, dtype: float64

**Вывод: Больше всего просрочек допускают клиенты со статусом `'Не женат / не замужем'` и `гражданский брак `.
Меньшее количество просрочек у категории `'вдовец / вдова'` 
Категории `'женат / замужем'`  и `'в разводе'` имеют почти равный процент просрочек.** 

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

`'total_income_category'` содержит сдедующие категории дохода:
- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

Количество просрочек по отношению к категории дохода

In [35]:
data.groupby('total_income_category')['debt'].sum().sort_values(ascending=False)

total_income_category
C    1353
B     354
D      21
A       2
E       2
Name: debt, dtype: int64

Общее количество займов по отношению к категории дохода

In [36]:
data.groupby('total_income_category')['debt'].count().sort_values(ascending=False)

total_income_category
C    15921
B     5014
D      349
A       25
E       22
Name: debt, dtype: int64

Сотношение просрочек к общему количеству займов

In [37]:
(data.groupby('total_income_category')['debt'].sum() /
data.groupby('total_income_category')['debt'].count()).sort_values(ascending=False)

total_income_category
E    0.090909
C    0.084982
A    0.080000
B    0.070602
D    0.060172
Name: debt, dtype: float64

**Вывод: выборка по категориям А и Е слишком мала для точного анализа. В остальных категорях нет прямой зависимости просрочек от дохода. Меньше всего просрочек у катигории D**

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

`'purpose_category'` содержит следующие категории цели кредита:
- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

Количество просрочек по отношению к категории цели кредита

In [38]:
data.groupby('purpose_category')['debt'].sum().sort_values(ascending=False)

purpose_category
операции с недвижимостью    780
операции с автомобилем      400
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

Общее количество займов по отношению к категории цели кредита

In [39]:
data.groupby('purpose_category')['debt'].count().sort_values(ascending=False)

purpose_category
операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: debt, dtype: int64

Сотношение просрочек к общему количеству займов

In [40]:
(data.groupby('purpose_category')['debt'].sum() /
data.groupby('purpose_category')['debt'].count()).sort_values(ascending=False)

purpose_category
операции с автомобилем      0.093480
получение образования       0.092528
проведение свадьбы          0.079118
операции с недвижимостью    0.072551
Name: debt, dtype: float64

**Вывод: Больше всего просрочек в категориях `'операции с автомобилем'`,`'получение образования'`** 

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

*В таблице пропуски были в двух столбцах это days_employed — общий трудовой стаж в днях и total_income — ежемесячный доход.
Возможно люди не указали эти данные по личным соображениям или в связи с отсутствием стажа или дохода* 

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ: заполнять пропуски средним значением допустимо в случае если распределение симметрично, унимодально и нет заметных выбросов т.к. в при таком распределении значения медианы моды и среднего будут находиться достаточно близко чтобы заполнение средним сильно повлияло на распределение. В остальных случаях предпочтительно заполнять пропуски медианой.* 

## Общий вывод.

После обработки и анализа полученных данных можно выделить несколько ярковыраженных категорий заемщиков по проценту просрочек.
При оценки зависимости просрочки от семейного положения были получены такие данные: 

Не женат / не замужем    9.7% \
гражданский брак         9.3% \
женат / замужем          7.5% \
в разводе                7% \
вдовец / вдова           6.6% \
\
Из данной таблицы можно выделить две категории заемщиков. \
Категории 'не женат / не замужем'  и 'гражданский брак' имеют самый большой процент просрочек. \
Категория 'вдовец / вдова' имеет самый малый процент просрочек. 

При оценки зависимости просрочек от количества детей у заемщиков было обнаруженно что выборка по многодетным заемщикам с разным количетвом детей слишком мала и все многодетные заемщики были объеденены в одну категорию.

многодетная семья    9.2% \
один ребенок         9.2% \
нет детей            7.5% \

По результатам данной таблицы можно выделить две группы заемщики с детьми и без детей. \
Заемщики с детьми допускают больше просрочек нежели заемщики без детей на 1.7%

Также были проанализированы заемщики по цели займа:

операции с автомобилем      9.3% \
получение образования       9.25% \
проведение свадьбы          7.9% \
операции с недвижимостью    7.25% 

Больше всего допускают просрочки по кредитам холостые люди и с детьми с доходом категории D (30000-50000 дохода). \
Самая большая доля просрочек по автокредитам, по кредитам на операции с недвижимостью процент просрочек меньше всего. 
